app/src/main/cpp/skyline/vfs/rom_filesystem.cpp

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2020 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#include "region_backing.h"
+#include "rom_filesystem.h"
+
+namespace skyline::vfs {
+    RomFileSystem::RomFileSystem(std::shared_ptr <Backing> backing) : FileSystem(), backing(backing) {
+        backing->Read(&header);
+
+        TraverseDirectory(0, "");
+    }
+
+    void RomFileSystem::TraverseFiles(u32 offset, std::string path) {
+        RomFsFileEntry entry{};
+
+        do {
+            backing->Read(&entry, header.fileMetaTableOffset + offset);
+
+            if (entry.nameSize) {
+                std::vector<char> name(entry.nameSize);
+                backing->Read(name.data(), header.fileMetaTableOffset + offset + sizeof(RomFsFileEntry), entry.nameSize);
```

Comment:
It might be better to read in the `fileMetaTable` into a `vector`, as these repeated backing lookups might be quite costly especially in a compressed format.

Reply:
I thought of that too but Backing reads are fairly cheap even then, constructing a romfs isn't done often either so it's not exactly performance sensitive